### PR TITLE
Add detailed contribution guidelines

### PR DESCRIPTION
**Pull request by sunandabalu**

### CODE REVIEW

File: .github/CODEOWNERS [hunk 1]
```diff
@@ -0,0 +1,4 @@
+# Users referenced in this file will automatically be requested as reviewers for PRs that modify the given paths.
+# See https://help.github.com/articles/about-code-owners/
+
+*                                                                                 @dotnet/dotnet-upgrade-assistant-admin
```
Comment (mjrousos) [inline .github/CODEOWNERS:4]: Super nit-pick, there seems to a be a lot of unnecessary white space on this line after the `*`

File: .github/ISSUE_TEMPLATE/10_design_proposal.md [hunk 2]
```diff
@@ -0,0 +1,49 @@
+---
```
Comment (mjrousos) [inline .github/ISSUE_TEMPLATE/10_design_proposal.md:1]: If we have a design proposal template, should we also have a bug report template? This one seems to be more about proposing something new but it doesn't cover the case of "this isn't working right" as well.

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -1,12 +1,80 @@
...
 For non-security related bugs, please [log a new issue](https://github.com/dotnet/upgrade-assistant/issues).
 
+## How to submit a PR 
+
+We are always happy to see PRs from community members both for bug fixes as well as new features.
+To help you be successful we've put together a few simple rules to follow when you prepare to contribute to our codebase:
+
+**Finding an issue to work on**
+
+  We've created a separate bucket of issues, which would be great candidates for community members to contribute to. We mark these issues with `help wanted` label. You can find all these issues [here](https://github.com/dotnet/upgrade-assistant/issues?q=is%3Aissue+is%3Aopen+label%3A%22help+wanted%22).
```
Comment (mjrousos) [inline CONTRIBUTING.md:18]:
```suggestion
  We've created a separate bucket of issues which would be great candidates for community members to contribute to. We mark these issues with the `help wanted` label. You can find all these issues [here](https://github.com/dotnet/upgrade-assistant/issues?q=is%3Aissue+is%3Aopen+label%3A%22help+wanted%22).
```

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -1,12 +1,80 @@
...
+  
+If there is some other area not included here where you want to contribute to, first open an issue to describe the problem you are trying to solve and state that you're willing to contribute a fix for it. We will then discuss in that issue, if this is an issue we would like addressed and the best approach to solve it.
+  
+**Before writing code**
+
+  This can save you a lot of time. We've seen PRs, where customers would solve an issue in a way, which either wouldn't fit into upgrade-assistant because of how it's designed or it would change upgrade-assistant in a way, which is not something we'd like to do. To avoid these situations, we encourage customers to discuss the preferred design with the team first. To do so, file a new `design proposal` issue, link to the issue you'd like to address and provide detailed information about how you'd like to solve a specific problem. We triage issues peridocially and it will not take long for a team member to engage with you on that proposal.
```
Comment (mjrousos) [inline CONTRIBUTING.md:26]:
```suggestion
  This can save you a lot of time. We've seen PRs where customers solve an issue in a way which either wouldn't fit into upgrade-assistant because of how it's designed or would change upgrade-assistant in a way which is not something we'd like to do. To avoid these situations, we encourage customers to discuss the preferred design with the team first. To do so, file a new `design proposal` issue, link to the issue you'd like to address and provide detailed information about how you'd like to solve a specific problem. We triage issues periodically and it will not take long for a team member to engage with you on that proposal.
```

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -1,12 +1,80 @@
...
+  
+**Before submitting the pull request**
+
+Before submitting a pull request, make sure that it checks the following requirements:
+
+- You find an existing issue with the "help-wanted" label or discuss with the team to agree on adding a new issue with that label
+- You post a high-level description of how it will be implemented, and receive a positive acknowledgement from the team before getting too committed to the approach or investing too much effort implementing it
+- You add test coverage following existing patterns within the codebase
+- Your code matches the existing syntax conventions within the codebase
+- Your PR is small, focused, and avoids making unrelated changes
```
Comment (mjrousos) [inline CONTRIBUTING.md:39]: Small is relative, so it might be easier to just say 'Your PR is focused and avoids making unrelated changes' since those criteria are a little easier to identify.

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -1,12 +1,80 @@
...
+  
+If your pull request contains any of the below, it's less likely to be merged.
```
Comment (mjrousos) [inline CONTRIBUTING.md:41]:
```suggestion
If your pull request contains any of the below, it's less likely to be merged:
```

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -1,12 +1,80 @@
...
+
+- Changes that break backward compatibility
+- Changes that are only wanted by one person/company. Changes need to benefit a large enough proportion of upgrade-assistant users.
```
Comment (mjrousos) [inline CONTRIBUTING.md:44]:
```suggestion
- Changes that are only wanted by one person/company. Changes need to benefit a large enough portion of upgrade-assistant users.
```

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -1,12 +1,80 @@
...
+
+### Submitting a pull request
+
+You will need to sign a [Contributor License Agreement](https://cla.dotnetfoundation.org/) when submitting your pull request. To complete the Contributor License Agreement (CLA), you will need to follow the instructions provided by the CLA bot when you send the pull request. This needs to only be done once for any .NET Foundation OSS project.
+
+If you don't know what a pull request is read this article: https://help.github.com/articles/using-pull-requests. Make sure the repository can build and all tests pass. Familiarize yourself with the project workflow and our coding conventions. The coding, style, and general engineering guidelines are published on the [Engineering guidelines](https://github.com/dotnet/aspnetcore/wiki/Engineering-guidelines) page.
```
Comment (mjrousos) [inline CONTRIBUTING.md:66]: Do these engineering guidelines apply to upgrade-assistant? I think some of them (External dependencies, Build numbers, Branch strategy, solutions filters, etc.) don't make sense in the context of upgrade-assistant.

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -1,12 +1,80 @@
...
+
+### Tests
+
+-  Tests need to be provided for every bug/feature that is completed.
+-  Tests only need to be present for issues that need to be verified by QA (for example, not tasks)
```
Comment (mjrousos) [inline CONTRIBUTING.md:71]: It isn't entirely clear to me what this means. What are 'tasks' and why don't they need QA? I could see PR related to docs or samples or things like that not needing tests but I'm not sure what the example given here is referring to.

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -1,12 +1,80 @@
...
+-  If there is a scenario that is far too hard to test there does not need to be a test for it.
```
Comment (mjrousos) [inline CONTRIBUTING.md:72]: This (and the following bullet) is another area that seems gray to me.  On the one hand, I think this bullet is true. But on the other hand, I think cases that are both well-designed and too hard to test are rare enough that this bullet isn't necessary.

I think most of upgrade-assistant should be unit testable. I think enough of it should be unit testable that it might be better to just have as-needed conversations in PRs that are difficult to test for some reason. Often, "far too hard to test" means "spaghetti code" or "not designed with testing in mind," both of which can be fixed without needing a "it's too hard to test" excuse for not having unit tests.

File: CONTRIBUTING.md [hunk 11]
```diff
@@ -1,12 +1,80 @@
...
+- "Too hard" is determined by the team as a whole.
+
+### Feedback
+
+Your pull request will now go through extensive checks by the subject matter experts on our team. Please be patient, while upgrade-assistant team gets through it. Update your pull request according to feedback until it is approved by one of the upgrade-assistant team members. Once the PR is approved, one of the upgrade-assistant team members will merge your PR into the repo.
```
Comment (mjrousos) [inline CONTRIBUTING.md:77]:
```suggestion
Your pull request will now go through extensive checks by the subject matter experts on our team. Please be patient while upgrade-assistant team gets through it. Update your pull request according to feedback until it is approved by one of the upgrade-assistant team members. Once the PR is approved, one of the upgrade-assistant team members will merge your PR into the repo.
```

**Review by mjrousos — CHANGES_REQUESTED** [review]
Looks good. I added a few comments and grammatical suggestions inline.

File: CONTRIBUTING.md (hunk 5 above)
Comment (sunandabalu) [reply CONTRIBUTING.md:39]: there is a line below to break down larger PRs into smaller chunks, so it makes sense to leave the term small here IMO

File: CONTRIBUTING.md (hunk 10 above)
Comment (sunandabalu) [reply CONTRIBUTING.md]: Removed this line

File: CONTRIBUTING.md (hunk 9 above)
Comment (sunandabalu) [reply CONTRIBUTING.md]: Removed this line

File: .github/CODEOWNERS (hunk 1 above)
Comment (sunandabalu) [reply .github/CODEOWNERS]: reduced some of the spaces :)

File: .github/ISSUE_TEMPLATE/10_design_proposal.md (hunk 2 above)
Comment (sunandabalu) [reply .github/ISSUE_TEMPLATE/10_design_proposal.md:1]: Added a bug report template.

File: CONTRIBUTING.md (hunk 8 above)
Comment (sunandabalu) [reply CONTRIBUTING.md:66]: Agree, not all sections apply to us. Maybe just a link to the coding-guidelines would work? https://github.com/dotnet/aspnetcore/wiki/Engineering-guidelines#coding-guidelines

File: CONTRIBUTING.md [hunk 12]
```diff
@@ -1,12 +1,78 @@
...
+- Changes that break backward compatibility
+- Changes that are only wanted by one person/company. Changes need to benefit a large enough portion of upgrade-assistant users.
+- Changes that add entirely new feature areas without prior agreement
+- Changes that are mostly about refactoring existing code or code style
+- Very large PRs that would take hours to review (remember, we're trying to help lots of people at once). For larger work areas, please discuss with us to find ways of breaking it down into smaller, incremental pieces that can go into separate PRs.
+
+**During pull request review**
+A core contributor will review your pull request and provide feedback. To ensure that there is not a large backlog of inactive PRs, the pull request will be marked as stale after two weeks of no activity. After another four days, it will be closed.
+
+### Resources to help you get started
```
Comment (KSchlobohm) [inline CONTRIBUTING.md:52]: sounds very similar to the ** Finding an issue to work on ** section. do we need both? or did I miss the intent?

**Review by cathysull — APPROVED** [review]
Looks good - one thought I had on the "Tests" section. Is it self-explanatory on how to include a test or is there an example we can point to? Just a thought but no need to address it right now if it's something we could add an example for.

File: CONTRIBUTING.md [hunk 13]
```diff
@@ -1,12 +1,78 @@
...
+
+### Submitting a pull request
+
+You will need to sign a [Contributor License Agreement](https://cla.dotnetfoundation.org/) when submitting your pull request. To complete the Contributor License Agreement (CLA), you will need to follow the instructions provided by the CLA bot when you send the pull request. This needs to only be done once for any .NET Foundation OSS project.
+
+If you don't know what a pull request is read this article: https://help.github.com/articles/using-pull-requests. Make sure the repository can build and all tests pass. Familiarize yourself with the project workflow and our coding conventions. For general coding guidelines, see [here](https://github.com/dotnet/aspnetcore/wiki/Engineering-guidelines#coding-guidelines).
+
+### Tests
+
+-  Tests need to be provided for every bug/feature that is completed.
```
Comment (twsouthwick) [inline CONTRIBUTING.md:70]: It may be good to point to an example. We've been making use of XUnit, AutoMock, and AutoFixture for testing/mocking/data generation. Maybe describe each of those with links

**Review by twsouthwick — COMMENTED** [review]
A few comments, but this looks good

File: CONTRIBUTING.md (hunk 12 above)
Comment (sunandabalu) [reply CONTRIBUTING.md:52]: The intent is to give a TLDR list of links. I think it will be useful to have this short list.

File: CONTRIBUTING.md [hunk 14]
```diff
@@ -1,12 +1,78 @@
...
+
+- You find an existing issue with the "help-wanted" label or discuss with the team to agree on adding a new issue with that label
+- You post a high-level description of how it will be implemented, and receive a positive acknowledgement from the team before getting too committed to the approach or investing too much effort implementing it
+- You add test coverage following existing patterns within the codebase
+- Your code matches the existing syntax conventions within the codebase
+- Your PR is small, focused, and avoids making unrelated changes
+  
+If your pull request contains any of the below, it's less likely to be merged:
+
+- Changes that break backward compatibility
```
[the remark this answers is not in the document]
Comment (sunandabalu) [reply CONTRIBUTING.md:43]: This are generic guidelines for contributing in any OSS repo, even though we may not be there yet its good to have it listed.

File: CONTRIBUTING.md (hunk 13 above)
Comment (sunandabalu) [reply CONTRIBUTING.md]: Added more details on tests in UA with some links. Do you know of alternative links to what I have linked?

File: CONTRIBUTING.md [hunk 15]
```diff
@@ -1,12 +1,78 @@
...
+
+### Submitting a pull request
+
+You will need to sign a [Contributor License Agreement](https://cla.dotnetfoundation.org/) when submitting your pull request. To complete the Contributor License Agreement (CLA), you will need to follow the instructions provided by the CLA bot when you send the pull request. This needs to only be done once for any .NET Foundation OSS project.
+
+If you don't know what a pull request is read this article: https://help.github.com/articles/using-pull-requests. Make sure the repository can build and all tests pass. Familiarize yourself with the project workflow and our coding conventions. For general coding guidelines, see [here](https://github.com/dotnet/aspnetcore/wiki/Engineering-guidelines#coding-guidelines).
+
+### Tests
+
+-  Tests need to be provided for every bug/feature that is completed. [Tests](/tests) in upgrade-assistant use [XUnit](https://docs.microsoft.com/en-us/dotnet/core/testing/unit-testing-with-dotnet-test), [AutoMock](https://autofac.org/apidoc/html/6C60CE20.htm), and [AutoFixture](https://github.com/AutoFixture/AutoFixture/wiki/Cheat-Sheet) for testing/mocking/data generation.
```
Comment (twsouthwick) [inline CONTRIBUTING.md:70]: I'd put each of them on their own line.

xunit: testing framework
automock; mocking
autofixture: data generation

Use this link for automock: https://autofaccn.readthedocs.io/en/latest/integration/moq.html

Comment (sunandabalu) [reply CONTRIBUTING.md]: Reworded.

**Comment by sunandabalu** [comment]
@mjrousos I have addressed the comments, could you PTAL and reset your vote if there are no more comments? :)

File: CONTRIBUTING.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (mjrousos) [reply CONTRIBUTING.md:43]: I don't think it's blocking, but I do agree with Taylor that at the moment we're not thinking about compat much so it probably doesn't need to be here (yet).

File: CONTRIBUTING.md [hunk 16]
```diff
@@ -1,12 +1,84 @@
...
+You will need to sign a [Contributor License Agreement](https://cla.dotnetfoundation.org/) when submitting your pull request. To complete the Contributor License Agreement (CLA), you will need to follow the instructions provided by the CLA bot when you send the pull request. This needs to only be done once for any .NET Foundation OSS project.
+
+If you don't know what a pull request is read this article: https://help.github.com/articles/using-pull-requests. Make sure the repository can build and all tests pass. Familiarize yourself with the project workflow and our coding conventions. For general coding guidelines, see [here](https://github.com/dotnet/aspnetcore/wiki/Engineering-guidelines#coding-guidelines).
+
+### Tests
+
+[Tests](/tests) in upgrade-assistant follow the following pattern:
+
+- Testing Framework used is [XUnit](https://docs.microsoft.com/en-us/dotnet/core/testing/unit-testing-with-dotnet-test).
+- Mocking Framework used is [AutoMock](https://autofaccn.readthedocs.io/en/latest/integration/moq.html).
```
Comment (mjrousos) [inline CONTRIBUTING.md:73]:
```suggestion
- Mocking Framework used is [Moq](https://github.com/Moq/moq4) (with [AutoMock](https://autofaccn.readthedocs.io/en/latest/integration/moq.html)).
```

File: .github/pull_request_template.md [hunk 17]
```diff
@@ -0,0 +1,21 @@
+
+<!-- Thank you for submitting a pull request to our repo. -->
+
+<!-- If this is your first PR in the upgrade-assistant repo, please run through the checklist
+below to ensure a smooth review and merge process for your PR. -->
+
+- [ ] You've read the [Contributor Guide](https://github.com/dotnet/upgrade-assistant/blob/main/CONTRIBUTING.md) and [Code of Conduct](https://github.com/dotnet/upgrade-assistant/blob/main/CODE-OF-CONDUCT.md).
```
Comment (mjrousos) [inline .github/pull_request_template.md:7]: These links can be shortened since they're in this same repo.

**Review by mjrousos — APPROVED** [review]
Looks good. I added a couple more comments inline but looks good overall!

File: .github/ISSUE_TEMPLATE/20_bug_report.md [hunk 18]
```diff
@@ -0,0 +1,25 @@
...
+Provide the log.txt with the error observed, along with project file(s) and / or if possible minimalistic repro project hosted in a GitHub repo.
+-->
+
+### Exceptions (if any)
+<!-- 
+Include the exception you get when facing this issue
+-->
+
+### Further technical details
+- Windows version and bitness info (32-bit or 64-bit)
```
[the remark this answers is not in the document]
Comment (twsouthwick) [reply .github/ISSUE_TEMPLATE/20_bug_report.md:24]: You could also ask for the output of `upgrade-assistant --version`

File: CONTRIBUTING.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (sunandabalu) [reply CONTRIBUTING.md:43]: Would regression be a better term to use here? Basically I want to make sure changes don't break any existing functionality.

**Comment by sunandabalu** [comment]
> Looks good - one thought I had on the "Tests" section. Is it self-explanatory on how to include a test or is there an example we can point to? Just a thought but no need to address it right now if it's something we could add an example for.

I have included the links to our existing tests, that can be used as an example for getting started on tests. I don't think a separate test sample is required.

File: CONTRIBUTING.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (sunandabalu) [reply CONTRIBUTING.md]: Re-worded to convey the point.
